### PR TITLE
[#5699] Specify field for instruction "on conflict … do ignore" 

### PR DESCRIPTION
**Pull request by fleroy84**
Suggestion of implementation for ticket #5699

### CODE REVIEW

**Comment by lukaseder** [comment]
Thank you very much for your suggestion. I'll review it shortly

File: jOOQ/src/main/java/org/jooq/InsertOnConflictDoActionStep.java [hunk 1]
```diff
@@ -59,11 +59,17 @@
  *
  * @author Lukas Eder
  */
-public interface InsertOnConflictDoUpdateStep<R extends Record> {
+public interface InsertOnConflictDoActionStep<R extends Record> {
```
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/InsertOnConflictDoActionStep.java]: We cannot rename types due to backwards compatibility constraints. It's OK if the type name doesn't fully match all of its features...

File: jOOQ/src/main/java/org/jooq/InsertOnConflictDoActionStep.java [hunk 2]
```diff
@@ -59,11 +59,17 @@
...
      */
     @Support({ POSTGRES_9_5 })
     InsertOnDuplicateSetStep<R> doUpdate();
+
+    /**
+     * Add the <code>DO IGNORE</code> clause.
+     */
+    @Support({ POSTGRES_9_5 })
+    InsertFinalStep<R> doIgnore();
 }
```
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/InsertOnConflictDoActionStep.java]: The actual syntax in PostgreSQL is actually `DO NOTHING`. I would prefer if we repeat that in our API as well

File: jOOQ/src/main/java/org/jooq/impl/InsertImpl.java [hunk 3]
```diff
@@ -580,6 +580,13 @@ public final InsertImpl defaultValues() {
...
     @Override
     public final InsertImpl doUpdate() {
+        onDuplicateKeyUpdate = true;
+        return this;
+    }
+
+    @Override
+    public final InsertImpl doIgnore() {
+        getDelegate().onDuplicateKeyIgnore(true);
         return this;
```
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/impl/InsertImpl.java]: I think that from a "principle of least surprise" point of view, we should now set `onDuplicateKeyUpdate` in both of these methods, and we should stop setting the flag in `onConflict()`

File: jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java [hunk 4]
```diff
@@ -78,12 +78,14 @@
...
 import org.jooq.exception.SQLDialectNotSupportedException;
+import org.jooq.tools.JooqLogger;
 
 /**
  * @author Lukas Eder
  */
 final class InsertQueryImpl<R extends Record> extends AbstractStoreQuery<R> implements InsertQuery<R> {
 
+    private static final JooqLogger LOGGER = JooqLogger.getLogger(InsertQueryImpl.class);
     private static final long        serialVersionUID = 4466005417945353842L;
```
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java]: Minor thing: Our logger is always called `log`

File: jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java [hunk 5]
```diff
@@ -278,10 +279,21 @@ else if (onDuplicateKeyIgnore) {
...
-                       .keyword("on conflict do nothing")
-                       .end(INSERT_ON_DUPLICATE_KEY_UPDATE);
+                    ctx.formatSeparator().start(INSERT_ON_DUPLICATE_KEY_UPDATE).keyword("on conflict");
+
+                    if (onConflict != null && onConflict.size() > 0) {
+                        if (onConflict.size() != 1) {
+                            LOGGER.info(
+                                    "ON DUPLICATE KEY IGNORE accept only one field for unique/exclusion constraint, " +
+                                            "let's take the first one on the list {}", onConflict.get(0).getName());
+                        }
```
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java]: Is this really correct? Can you point me to the docs where this is specified? I would imagine that we should be able to list the columns of a composite key here...

File: jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java [hunk 6]
```diff
@@ -278,10 +279,21 @@ else if (onDuplicateKeyIgnore) {
...
+                        ctx.sql(" (");
+                        boolean qualify = ctx.qualify();
+                        ctx.qualify(false).visit(onConflict.get(0)).qualify(qualify);
+                        ctx.sql(") ");
+                    }
+
+                    ctx.sql(" do nothing").end(INSERT_ON_DUPLICATE_KEY_UPDATE);
```
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java]: Keywords are passed to the `ctx` using `ctx.keyword()`, i.e. this should be:

```java
ctx.sql(' ').keyword("do nothing")...
```

File: jOOQ/src/main/java/org/jooq/impl/InsertImpl.java (hunk 3 above)
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/impl/InsertImpl.java]: I see, you've done the latter in a separate commit:
https://github.com/jOOQ/jOOQ/pull/5702/commits/8eb788082d57fd34b6cf82c3833e253c86f47ca7

File: jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java (hunk 6 above)
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java]: OK, that's been done in a separate commit: https://github.com/jOOQ/jOOQ/pull/5702/commits/b75cf702a8ee272fc7e62adf25c0cf463fb0e131

**Comment by fleroy84** [comment]
You suggest to add a method “onConflictDoNothing()”. This method will be duplicate of “onDuplicateKeyIgnore()”, no ?
 Or may be you want this method to be coherent with postgresql instruction ?

```
getDslContext().insertInto[myTable])).set([myRecord]).onDuplicateKeyIgnore());

getDslContext().insertInto[myTable])).set([myRecord]).onConflictDoNothing());
```

This 2 lines of code, will do the same thing, right ?

File: jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java (hunk 5 above)
Comment (fleroy84) [inline jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java]: You're right, an error has been made in my test case ! I will fix it.

**Comment by lukaseder** [comment]
> These 3 lines of code, will do the same thing, right ?

Yes, they'll do the same thing.

I'll merge your updated suggestion and will further comment / fix things later on as the 3.9 release is due this week. Thanks again for your contribution, this is greatly appreciated!
